### ISSUE TITLE
Is it possible to develop & run the project locally without deploying on Heroku?

### ISSUE THREAD

**Issue by bhavyakaria** [google/wikiloop-battlefield#240]
@xinbenlv

**Comment by xinbenlv** [comment]
Yes, definitely. See https://github.com/google/wikiloop-battlefield#deploy-a-dev-instance-on-heroku, and let me know if you have any more questions

**Comment by bhavyakaria** [comment]
Hi @xinbenlv,

I did all the steps present in the README and after running the ```npm run dev``` command, I'm getting the following error

**Steps Followed**
1. npm install
2. npm run dev

```
/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/connection_config.js:229
      database: url.pathname.substr(1)
                             ^
TypeError: Cannot read property 'substr' of null
    at Function.parseUrl (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/connection_config.js:229:30)
    at new PoolConfig (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/pool_config.js:8:34)
    at Object.exports.createPool (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/index.js:20:29)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/server/routes/score.ts:24:20)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Module.m._compile (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:837:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:840:12)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:690:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/server/index.ts:45:1)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Module.m._compile (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:837:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:840:12)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:829:12)
    at main (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/bin.ts:226:14)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/bin.ts:485:3)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! wikiloop-battlefield@4.0.1-beta dev:exec: `ts-node -r tsconfig-paths/register --project tsconfig.json ./server/index.ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the wikiloop-battlefield@4.0.1-beta dev:exec script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bhavyakaria/.npm/_logs/2020-05-10T17_50_00_304Z-debug.log
```

And my .env file looks like this,
```
# this file was created automatically by heroku-config
# This is our GAID for dev
GA_WLBF_ID_CLIENT=UA-122681777-3
GA_WLBF_ID_API=UA-122681777-3

HOST="0.0.0.0"
NPM_CONFIG_PRODUCTION=false

# We intentionally leave this empty.
MONGODB_URI=

# We will work with STIKI Author to work out a public solution
STIKI_MYSQL=

# Get it here https://meta.wikimedia.org/wiki/Special:OAuthConsumerRegistration/list
# Instructions https://www.mediawiki.org/wiki/OAuth/For_Developers
# for debug https://www.mediawiki.org/wiki/OAuth/Owner-only_consumers
MEDIAWIKI_CONSUMER_KEY=
MEDIAWIKI_CONSUMER_SECRET=

PUBLIC_HOST=localhost:3000
DIRECT_REVERT=1
```

Due to some issues, I can't do Deploy to Heroku. Have I missed any step?

**Comment by xinbenlv** [comment]
Right, you will need to have a `MONGODB_URI=` value. You will need to have a MongoDB server. An easy solution is to launch a free on on Heroku. but if you found not inconvenient, feel free to use our dev one:
`mongodb://heroku_2dmqvqnm:h09qmsls5345p05e30mk1n39vi@ds261828.mlab.com:61828/heroku_2dmqvqnm`

**Comment by bhavyakaria** [comment]
I have added the provided URI to ```MONGODB_URI=```. But I'm still getting the same error.

```
# this file was created automatically by heroku-config
# This is our GAID for dev
GA_WLBF_ID_CLIENT=UA-122681777-3
GA_WLBF_ID_API=UA-122681777-3

HOST="0.0.0.0"
NPM_CONFIG_PRODUCTION=false

# We intentionally leave this empty.
MONGODB_URI=mongodb://heroku_2dmqvqnm:h09qmsls5345p05e30mk1n39vi@ds261828.mlab.com:61828/heroku_2dmqvqnm

# We will work with STIKI Author to work out a public solution
STIKI_MYSQL=

# Get it here https://meta.wikimedia.org/wiki/Special:OAuthConsumerRegistration/list
# Instructions https://www.mediawiki.org/wiki/OAuth/For_Developers
# for debug https://www.mediawiki.org/wiki/OAuth/Owner-only_consumers
MEDIAWIKI_CONSUMER_KEY=
MEDIAWIKI_CONSUMER_SECRET=

PUBLIC_HOST=localhost:3000
DIRECT_REVERT=1
```

**Comment by xinbenlv** [comment]
Can you show me the new error?

**Comment by bhavyakaria** [comment]
It's the same error as commented above.

```
❯ npm run dev

> wikiloop-battlefield@4.0.1-beta dev /Users/bhavyakaria/wade_watts/web/wikiloop-battlefield
> cross-env NODE_ENV=development nodemon

[nodemon] 1.19.4
[nodemon] to restart at any time, enter `rs`
[nodemon] watching dir(s): server/**/* cron mailer/**/* shared/**/*
[nodemon] watching extensions: ts
[nodemon] starting `npm run dev:exec`

> wikiloop-battlefield@4.0.1-beta dev:exec /Users/bhavyakaria/wade_watts/web/wikiloop-battlefield
> ts-node -r tsconfig-paths/register --project tsconfig.json ./server/index.ts


/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/connection_config.js:229
      database: url.pathname.substr(1)
                             ^
TypeError: Cannot read property 'substr' of null
    at Function.parseUrl (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/connection_config.js:229:30)
    at new PoolConfig (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/lib/pool_config.js:8:34)
    at Object.exports.createPool (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/mysql2/index.js:20:29)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/server/routes/score.ts:24:20)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Module.m._compile (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:837:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:840:12)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:690:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/server/index.ts:45:1)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Module.m._compile (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:837:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/index.ts:840:12)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:829:12)
    at main (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/bin.ts:226:14)
    at Object.<anonymous> (/Users/bhavyakaria/wade_watts/web/wikiloop-battlefield/node_modules/ts-node/src/bin.ts:485:3)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! wikiloop-battlefield@4.0.1-beta dev:exec: `ts-node -r tsconfig-paths/register --project tsconfig.json ./server/index.ts`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the wikiloop-battlefield@4.0.1-beta dev:exec script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bhavyakaria/.npm/_logs/2020-05-10T18_25_10_732Z-debug.log
[nodemon] app crashed - waiting for file changes before starting...
```

**Comment by xinbenlv** [comment]
That should work, try add the quotes?

```
MONGODB_URI="mongodb://heroku_2dmqvqnm:h09qmsls5345p05e30mk1n39vi@ds261828.mlab.com:61828/heroku_2dmqvqnm"
````

**Comment by xinbenlv** [comment]
Hold on, I might be wrong. Do you mind move the conversation to https://discord.gg/nUuCbB so we can use it as instant messenger?

**Comment by xinbenlv** [comment]
Fixed by f7ae365e7382f0c1d0f7d799083ca0a01182fa89
